pkg/pillar/types/global.go

```diff
@@ -103,312 +110,367 @@ func (gs *GlobalStatus) UpdateItemValuesFromGlobalConfig(gc GlobalConfig) {
+type ConfigItemValueMap struct {
+	// GlobalSettings key is the agent name.
+	GlobalSettings map[string]ConfigItemValue
+	// AgentSettings key is the agent name. The inner map's key is the per-agent setting name.
+	AgentSettings map[string]map[string]ConfigItemValue
+}
+
+func getValueWithDefault(key string, agentName string, agentSetting bool, globalSetting bool,
+	config *ConfigItemValueMap, specMap *ConfigItemSpecMap) ConfigItemValue {
+	var okVal bool
+	var okSpec bool
+	var val ConfigItemValue
+	var spec ConfigItemSpec
+	if globalSetting {
+		val, okVal = config.GlobalSettings[key]
+		spec, okSpec = specMap.GlobalSettings[key]
+	}
+	if agentSetting {
+		agent, ok := config.AgentSettings[agentName]
+		if ok {
+			val, okVal = agent[key]
+		}
+		specM, ok := specMap.AgentSettings[agentName]
+		if ok {
+			spec, okSpec = specM[key]
+		}
+	}
+	if okVal {
+		return val
+	} else if okSpec {
+		return ParseSpecToValue(spec)
+	} else {
+		return ConfigItemValue{
+			IntValue:      0,
+			BoolValue:     false,
+			TriStateValue: TS_NONE,
+			StringValue:   "",
+		}
 	}
+}
 
-	if newgc.Dom0MinDiskUsagePercent == 0 {
-		newgc.Dom0MinDiskUsagePercent =
-			GlobalConfigDefaults.Dom0MinDiskUsagePercent
+func getValue(key string, agentName string, agentSetting bool, globalSetting bool,
+	config *ConfigItemValueMap) ConfigItemValue {
+	var val ConfigItemValue
+	var ok bool
+	if globalSetting {
+		val, ok = config.GlobalSettings[key]
+	}
+	if agentSetting {
+		val, ok = config.AgentSettings[agentName][key]
+	}
+	if ok {
+		return val
+	} else {
+		return ConfigItemValue{
+			IntValue:      0,
+			BoolValue:     false,
+			TriStateValue: TS_NONE,
+			StringValue:   "",
+		}
 	}
-	return newgc
+}
+func GetAgentValueWithDefault(key string, agentName string, config *ConfigItemValueMap, specMap *ConfigItemSpecMap) ConfigItemValue {
+	return getValueWithDefault(key, agentName, true, false, config, specMap)
 }
 
-// We enforce that timers are not below these values
-var GlobalConfigMinimums = GlobalConfig{
-	ConfigInterval:          5,
-	MetricInterval:          5,
-	ResetIfCloudGoneTime:    120,
-	FallbackIfCloudGoneTime: 60,
-	MintimeUpdateSuccess:    30,
-
-	NetworkGeoRedoTime:        60,
-	NetworkGeoRetryTime:       5,
-	NetworkTestDuration:       10,  // Wait for DHCP client
-	NetworkTestInterval:       300, // 5 minutes
-	NetworkTestBetterInterval: 0,   // Disabled
-
-	StaleConfigTime:         0, // Don't use stale config
-	DownloadGCTime:          60,
-	VdiskGCTime:             60,
-	DownloadRetryTime:       60,
-	DomainBootRetryTime:     10,
-	Dom0MinDiskUsagePercent: 20,
-	RktGCGracePeriod:        600,
+func GetAgentValue(key string, agentName string, config *ConfigItemValueMap) ConfigItemValue {
+	return getValue(key, agentName, true, false, config)
 }
 
-func EnforceGlobalConfigMinimums(newgc GlobalConfig) GlobalConfig {
+func GetGlobalValueWithDefault(key string, config *ConfigItemValueMap, specMap *ConfigItemSpecMap) ConfigItemValue {
```

Comment:
Why do u need 2 different functions?
   GetGlobalValueWithDefault
   GetGlobalValue

These are better as methods of ConfigItemSpecMap - Have those 6 methods described in comment above.